chore(batch-exports): Add batch export monitoring workflow

rossgray opened the pull request:
## Problem

In the past we've had some issues with events missing from certain batch exports and lack any reliable monitoring for this.

## Changes

Add a new workflow in Temporal for monitoring the events exported by a given team's batch export.

This workflow is quite specific at the moment, in that it just checks 5 min batch exports (since these are the only ones using the `recent_events` table). It checks for a previous hour's worth of events (based on `inserted_at`) and stores the event count in the `batch_export_runs` table, so we can later reconcile this against `records_completed`.

I have left a few TODOs in the code for things I'm unsure about.

## Does this work well for both Cloud and self-hosted?

it doesn't have an impact

## How did you test this code?

Added tests

rossgray committed:
WIP

rossgray committed:
Updates

rossgray committed:
Finalise things

rossgray commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+    """
+
+    # TODO - make this a list of team ids or single?
+    # or maybe a (list of) batch export id(s) instead?
```
maybe easier to use a batch export id instead in case the team has more than one set up?

tomasfarias replied:
I think it would be better to take a `batch_export_id` :thinking: But also fine with trying `team_id` and see how that works in practice if you are not convinced yet (I am not, maybe the solution is both...)

rossgray replied:
maybe I'll change this to take a `batch_export_id` for now since it makes some of the logic simpler and then can extend this to take a `team_id` if useful in future

tomasfarias replied:
Sounds good :+1: Thank you.

tomasfarias commented on posthog/batch_exports/models.py:
```diff
@@ -106,6 +106,13 @@ class Status(models.TextChoices):
     records_total_count = models.IntegerField(
         null=True, help_text="The total count of records that should be exported in this BatchExportRun."
     )
+    expected_records_count = models.IntegerField(
+        null=True,
+        help_text=(
+            "The total count of records that we expect to be exported in this BatchExportRun. "
+            "This is updated automatically post-batch export by a monitoring job."
+        ),
+    )
```
We can just use `records_total_count` and save the migration. It's essentially the same field, and it's currently not in use. Feel free to update the `help_text`.

rossgray replied:
ah nice, will do

tomasfarias commented on posthog/batch_exports/service.py:
```diff
@@ -794,3 +794,29 @@ async def aupdate_batch_export_backfill_status(backfill_id: UUID, status: str) -
+        batch_export_id=batch_export_id,
+        data_interval_start=interval_start,
+        data_interval_end=interval_end,
+    ).aupdate(expected_records_count=count)
```
Again, I'd just use the existing `records_total_count` here.

tomasfarias commented on posthog/batch_exports/sql.py:
```diff
@@ -318,3 +318,22 @@
+    interval_start + INTERVAL {interval} AS interval_end,
+    COUNT(*) as total_count
+FROM
+    events_batch_export_recent(
```
`events_batch_export_recent` should only be used with 5 minute batch exports. Other batch exports should use `events_batch_export_unbounded` or `events_batch_export`.

tomasfarias replied:
For the 5 minute batch export case though, the query looks fine :+1:

tomasfarias commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+    if len(models) > 1:
+        activity.logger.warning("More than one active events batch export found; using first one...")
+    model = models[0]
+    if model.interval_time_delta != dt.timedelta(minutes=5):
+        raise NoValidBatchExportsFoundError(
+            "Only batch exports with interval of 5 minutes are supported for monitoring at this time."
+        )
```
Ah, good, we are only doing 5 min. Then query is good :+1: Sorry, got to this part later.

tomasfarias commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+    models = await aget_active_event_batch_exports(team_id)
+    if len(models) == 0:
+        raise NoActiveBatchExportsFoundError(team_id)
+    if len(models) > 1:
+        activity.logger.warning("More than one active events batch export found; using first one...")
```
Perhaps we should make sure we use the latest one, as in the one created last.

tomasfarias replied:
If anything, just to be consistent and not rely on database whims.

rossgray replied:
Thanks. Going to use `batch_export_id` as an input so shouldn't have to worry about this

tomasfarias commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+
+        response = await client.read_query(query, query_params)
+        results = []
+        for line in response.decode("utf-8").splitlines():
```
I'm being nitpicky.

rossgray replied:
yeah ok, let's see if we need this someplace else before abstracting 👍

tomasfarias commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+    @workflow.run
+    async def run(self, inputs: BatchExportMonitoringInputs):
+        """Workflow implementation to monitor batch exports for a given team."""
+        # TODO - check if this is the right way to do logging since there seems to be a few different ways
+        workflow.logger.info("Starting batch exports monitoring workflow for team %s", inputs.team_id)
```
Most of our logging setup is to use kafka, ingest logs into ClickHouse, and display logs in the PostHog UI. But doing that at the Workflow code is impossible, so this is fine for now.

tomasfarias commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+            inputs.team_id,
+            start_to_close_timeout=dt.timedelta(minutes=1),
+            retry_policy=RetryPolicy(
+                maximum_attempts=3,
```
Shouldn't we retry forever here? Why would this ever fail (assuming the batch export exists, which seems to be covered)?

rossgray replied:
yeah good point 👍

tomasfarias commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+                include_events=batch_export_details.include_events,
+            ),
+            start_to_close_timeout=dt.timedelta(hours=1),
+            retry_policy=RetryPolicy(maximum_attempts=3, initial_interval=dt.timedelta(seconds=20)),
```
We probably want to exponentially backoff on retrying. As this will be hitting ClickHouse, the most common error will be a `TOO_MANY_SIMULTANEOUS_QUERIES` error from ClickHouse. This workflow should make sure it's not putting too much load in ClickHouse by being very generous with the retry interval when retrying.

rossgray replied:
Sounds good, although I think there is a default `backoff_coefficient` of 2 so currently this will retry after 20s, then again after 40s, then finally after 80s. Do you think it should be increased more?

tomasfarias replied:
Oh yeah, you are correct.

tomasfarias commented on posthog/temporal/batch_exports/monitoring.py:
```diff
@@ -0,0 +1,220 @@
+                initial_interval=dt.timedelta(seconds=20),
+                non_retryable_error_types=["NoActiveBatchExportsFoundError"],
+            ),
+            heartbeat_timeout=dt.timedelta(minutes=1),
```
This activity does not heartbeat, so having a heartbeat timeout is pointless.

tomasfarias replied:
Same thing with other activities: I'd add heartbeating (best) or remove the timeout as it is confusing (sufficient).

rossgray replied:
ah yes, oops, I just copy/pasted this from other jobs without considering that. Will look into adding a hearbeat

tomasfarias reviewed (changes requested):
Solution and implementation look really good. I have a handful of comments to address scattered all over, but nothing critical.

rossgray committed:
Use batch_export_id as input

rossgray committed:
Use records_total_count column

rossgray committed:
Add heartbeat

rossgray commented:
@tomasfarias thanks for the review! Think I've addressed all your feedback. Going to spend some more time investigating what's going on with this test

tomasfarias reviewed (approved):
I think PR looks great. Once tests are passing this is good to merge! Good work!

rossgray committed:
Fix tests

rossgray committed:
Add imports
